junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/JupiterTestDescriptor.java

```diff
@@ -182,15 +181,32 @@ public static ExecutionMode toExecutionMode(org.junit.jupiter.api.parallel.Execu
 		throw new JUnitException("Unknown ExecutionMode: " + mode);
 	}
 
-	Set<ExclusiveResource> getExclusiveResourcesFromAnnotation(AnnotatedElement element) {
+	protected List<ResourceLock> getResourceLocks() {
+		return Collections.emptyList();
+	}
+
+	protected Set<ExclusiveResource> collectExclusiveResourcesFromHierarchy() {
+		Set<ExclusiveResource> resources = mapResourceLocksForTarget(getResourceLocks(), ResourceLockTarget.SELF);
+
+		Optional<TestDescriptor> nextParent = getParent();
+		while (nextParent.isPresent()) {
```

Comment:
IIUC `@ResourceLock(target = CHILDREN)` should only cause them to be applied to _direct_ children.

Reply:
> IIUC `@ResourceLock(target = CHILDREN)` should only cause them to be applied to _direct_ children.

Is there any reason for this?  Is it harmful in someway for it to not be just direct children?  e.g. in our project our tests inherit from base classes scattered throughout our packages and those base classes tend to inherit from a single source parent.  If I have a resource that I want read locked for everything (main server component for example) it makes more sense to me to have the lock specified in one place instead of having to remember to add it to all the base classes

Reply:
@Squibbles This is unrelated to inheritance in class hierarchies. The `@ResourceLock` annotation is inherited to subclasses already. However, the way resource locks are applied is that if a node acquires a lock, its children already hold it while being executed. Therefore it doesn't make sense to recursively declare it for all descendants.